Library/Homebrew/os/mac/sdk.rb

```diff
@@ -25,14 +25,44 @@ class BaseSDKLocator
       class NoSDKError < StandardError; end
 
       def sdk_for(v)
-        path = sdk_paths[v]
-        raise NoSDKError if path.nil?
+        sdk = all_sdks.find { |s| s.version == v }
+        raise NoSDKError if sdk.nil?
 
-        SDK.new v, path, source
+        sdk
       end
 
       def all_sdks
-        sdk_paths.map { |v, p| SDK.new v, p, source }
+        @all_sdks ||= begin
+          # Bail out if there is no SDK prefix at all
+          if File.directory? sdk_prefix
+            sdks = []
+
+            Dir[File.join(sdk_prefix, "MacOSX*.sdk")].each do |sdk_path|
+              next unless sdk_path.match?(/MacOSX\d+\.\d+\.sdk$/)
+              next unless (version = read_sdk_version(Pathname.new(sdk_path)))
+
+              sdks << SDK.new(OS::Mac::Version.new(version), sdk_path, source)
+            end
+
+            # Use unversioned SDK path on Big Sur to avoid issues such as:
+            # https://github.com/Homebrew/homebrew-core/issues/67075
+            # This creates an entry in `paths` whose key is the OS major version
+            sdk_path = Pathname.new("#{sdk_prefix}/MacOSX.sdk")
+            if (version = read_sdk_version(sdk_path))
+              version = OS::Mac::Version.new(version)
+              if version >= :big_sur
+                # Clear all other previous SDKs for this version
+                sdks.reject! { |s| s.version == version }
```

Comment:
Could this array not be populated until after instead of populating then clearing it?